Add telemetry metrics related to client side aggregation (#11296)
DataDog/documentation: content/en/developers/dogstatsd/high_throughput.md

: **Metric type**: count<br>
The number of metrics dropped because the internal receiving channel is full (only when using `WithChannelMode()`). Starting with version `3.6.0` of the Go client.

`datadog.dogstatsd.client.aggregated_context`
: **Metric type**: count<br>
The number of contexts aggregated when client side aggregation is enabled. Starting with version `4.3.0`.

`datadog.dogstatsd.client.aggregated_context_by_type`
: **Metric type**: count<br>
The number of contexts aggregated by type when client side aggregation is enabled. Starting with version `4.3.0`.

`datadog.dogstatsd.client.metrics_by_type`
: **Metric type**: count<br>
The number of metrics by type sent to the DogStatsD client by your application (before sampling). Starting with version `4.3.0`.

> 

Reviewer: Those are behind `WithDevMode` ([here](https://github.com/DataDog/datadog-go/blob/master/statsd/telemetry.go#L141)) option until the next major. Also we should mentionned the tag `metrics_type` so user know how to use it.

> The number of contexts aggregated by type when client side aggregation is enabled. Starting with version `4.3.0`. The metric is enabled by default starting `5.0.0` but requires `WithDevMode` for the versions before or equal to `4.8.1`. The metric is tagged with either `metrics_type:gauge`, `metrics_type:count`, `metrics_type:set`, `metrics_type:timing`, `metrics_type:histogram` nor `metrics_type:distribution`.

Reviewer: We might release a new bug fix on v4 someday so I would just say:

```suggestion
The number of contexts aggregated by type when client side aggregation is enabled. Starting with version `4.3.0`. The metric is enabled by default starting `v5.0.0` but requires `WithDevMode()` option for the previous versions.
```

Also are we planning on blocking this commit until the release of both Java and Go new major ?

Reviewer 2: Also, let's be consistent and use `vX.Y.Z` or `X.Y.Z` everywhere, we currently have a bit of a mix. I think I like the `v` prepended because that's how we tag the repositories.

Author: Outdated as https://github.com/DataDog/documentation/pull/11900/files describes the same metrics. Changed removed.

: **Metric type**: count<br>
The number of datagrams dropped because the DogStatsD client queue was full.

`datadog.dogstatsd.client.aggregated_context`
: **Metric type**: count<br>
The number of contexts aggregated when client side aggregation is enabled. Starting with version `v2.11.0`.

`datadog.dogstatsd.client.aggregated_context_by_type`
: **Metric type**: count<br>

> The number of contexts aggregated by type when client side aggregation is enabled. Starting with version `v2.13.0`. The metric is enabled by default starting `v3.0.0` but requires `enableDevMode(true)` for `v2.13.0`. The metric is tagged with either `metrics_type:gauge`, `metrics_type:count`, `metrics_type:set`, `metrics_type:histogram` nor `metrics_type:distribution`.

Reviewer 3: ```suggestion
The number of contexts aggregated by type when client side aggregation is enabled. Starting with version `v2.13.0`. The metric is enabled by default starting `v3.0.0` but requires `enableDevMode(true)` for `v2.13.0`. The metric is tagged with either `metrics_type:gauge`, `metrics_type:count`, `metrics_type:set`, `metrics_type:histogram` or `metrics_type:distribution`.
```

`datadog.dogstatsd.client.metrics_by_type`
: **Metric type**: count<br>

> The number of metrics sent to the DogStatsD client by your application tagged by type (before sampling). Starting with version `v2.13.0` when `enableDevMode(true)` is used and by default starting `v3.0.0`. The metric is tagged with either `metrics_type:gauge`, `metrics_type:count`, `metrics_type:set`, `metrics_type:histogram` nor `metrics_type:distribution`.

Reviewer 3: ```suggestion
The number of metrics sent to the DogStatsD client by your application tagged by type (before sampling). Starting with version `v2.13.0` when `enableDevMode(true)` is used and by default starting `v3.0.0`. The metric is tagged with either `metrics_type:gauge`, `metrics_type:count`, `metrics_type:set`, `metrics_type:histogram` or `metrics_type:distribution`.
```

> The number of metrics by type sent to the DogStatsD client by your application (before sampling). Starting with version `v2.13.0`. The metric is enabled by default starting `v3.0.0` but requires `enableDevMode(true)` for `v2.13.0`. The metric is tagged with either `metrics_type:gauge`, `metrics_type:count`, `metrics_type:set`, `metrics_type:histogram` nor `metrics_type:distribution`.

Reviewer: What about:

```suggestion
The number of metrics sent to the DogStatsD client by your application tagged by type (before sampling). Starting with version `v2.13.0` when `enableDevMode(true)` is used and by default starting `v3.0.0`. The metric is tagged with either `metrics_type:gauge`, `metrics_type:count`, `metrics_type:set`, `metrics_type:histogram` nor `metrics_type:distribution`.
```

Also do the C# client support timing metrics ? Since the interface will list all tags I don't know if it's necessary to list all metric types. WDYT ?

Author: This section is about the java client.

Reviewer: Well, same question for java then

Author: cc @vickenty

Reviewer 3: Java client does not have timing metrics. I think we can mention only the tag key (`metrics_type`) without listing every possible metric type here.
